cmd/charmdelete/main.go

```diff
@@ -0,0 +1,163 @@
+// Copyright 2014-2017 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+package main
+
+import (
+	"flag"
+	"fmt"
+	"os"
+	"path/filepath"
+
+	"github.com/juju/loggo"
+	"gopkg.in/errgo.v1"
+	"gopkg.in/mgo.v2"
+
+	"gopkg.in/juju/charmstore.v5-unstable/config"
+	"gopkg.in/juju/charmstore.v5-unstable/elasticsearch"
+	"gopkg.in/juju/charmstore.v5-unstable/internal/charmstore"
+	"gopkg.in/juju/charmstore.v5-unstable/internal/mongodoc"
+	"gopkg.in/mgo.v2/bson"
+)
+
+var logger = loggo.GetLogger("charmdelete")
+
+var (
+	index             = flag.String("index", "cs", "Name of index to charmDelete.")
+	loggingConfig     = flag.String("logging-config", "", "specify log levels for modules e.g. <root>=TRACE")
+	user              = flag.String("user", "", "Delete all charms for a user. Do not mix with charmMatch.")
+	charmMatch        = flag.String("charm-match", "", "Delete all charms maching this expression. Do not mix with user.")
+	dryrun            = flag.Bool("dry-run", false, "Don't actually delete; just print them.")
+	verbose           = flag.Bool("verbose", false, "")
+	deletePromulgated = flag.Bool("delete-promulgated", false, "Delete a charm even if it is promulgated.")
+)
+
+func main() {
+	flag.Usage = func() {
+		fmt.Fprintf(os.Stderr, "usage: %s [options] <config path>\n", filepath.Base(os.Args[0]))
+		flag.PrintDefaults()
+		os.Exit(2)
+	}
+	flag.Parse()
+	if flag.NArg() != 1 {
+		flag.Usage()
+	}
+	if *loggingConfig != "" {
+		if err := loggo.ConfigureLoggers(*loggingConfig); err != nil {
+			fmt.Fprintf(os.Stderr, "cannot configure loggers: %v", err)
+			os.Exit(1)
+		}
+	}
+	if err := run(flag.Arg(0)); err != nil {
+		logger.Errorf("cannot run: %v", err)
+		os.Exit(1)
+	}
+}
+
+func run(confPath string) error {
+	logger.Debugf("reading config file %q", confPath)
+	conf, err := config.Read(confPath)
+	if err != nil {
+		return errgo.Notef(err, "cannot read config file %q", confPath)
+	}
+	if conf.ESAddr == "" {
+		return errgo.Newf("no elasticsearch-addr specified in config file %q", confPath)
+	}
+	si := &charmstore.SearchIndex{
+		Database: &elasticsearch.Database{
+			conf.ESAddr,
+		},
+		Index: *index,
+	}
+	session, err := mgo.Dial(conf.MongoURL)
+	if err != nil {
+		return errgo.Notef(err, "cannot dial mongo at %q", conf.MongoURL)
+	}
+	defer session.Close()
+	db := session.DB("juju")
+
+	pool, err := charmstore.NewPool(db, si, nil, charmstore.ServerParams{})
+	if err != nil {
+		return errgo.Notef(err, "cannot create a new store")
+	}
+	store := pool.Store()
+	defer store.Close()
+	entities := store.DB.Entities()
+
+	query := bson.D{{"user", *user}}
+	if *user == "" {
+		fmt.Printf("using match _id $regex %s\n", *charmMatch)
+		query = bson.D{{"_id", bson.D{{"$regex", *charmMatch}}}}
```

Comment:
Because this could match charms the admin doesn't intend it should probably print a list of ids that match and then a confirmation step. I see that there is a dryrun flag but it's not the default for this command, maybe it should be?